homeassistant/components/device_tracker/owntracks.py

```diff
@@ -60,47 +76,111 @@ def owntracks_event_update(topic, payload, qos):
             data = json.loads(payload)
         except ValueError:
             # If invalid JSON
-            logging.getLogger(__name__).error(
+            _LOGGER.error(
                 'Unable to parse payload as JSON: %s', payload)
             return
 
         if not isinstance(data, dict) or data.get('_type') != 'transition':
             return
 
-        # check if in "home" fence or other zone
-        location = ''
+        dev_id, host_name = _parse_topic(topic)
+
+        if data['desc'].lower() == 'home':
+            location = STATE_HOME
+        else:
+            location = data['desc']
+
+        kwargs = {
+            'dev_id': dev_id,
+            'host_name': host_name,
+            'gps': (data['lat'], data['lon'])
+        }
+
+        if 'acc' in data:
+            kwargs['gps_accuracy'] = data['acc']
+
         if data['event'] == 'enter':
+            zone = hass.states.get("zone.{}".format(location))
+            LOCK.acquire()
+
+            if zone is not None:
+                kwargs['location_name'] = location
+                _block_updates(dev_id, location)
+
+                if data['t'] == 'b':
+                    # For beacon events - the zone location is more
+                    # accurate than gps
+                    kwargs['gps'] = (
+                        zone.attributes['latitude'],
+                        zone.attributes['longitude'])
+                    kwargs['gps_accuracy'] = 1
 
-            if data['desc'].lower() == 'home':
-                location = STATE_HOME
-            else:
-                location = data['desc']
+            see(**kwargs)
+            LOCK.release()
+
+            # Block location update while we're in a region
 
         elif data['event'] == 'leave':
-            location = STATE_NOT_HOME
+            if not _valid_leave(dev_id, location):
+                return
+
+            see(**kwargs)
+
+            _block_updates(dev_id, '')
```

Comment:
If you move to a dict approach, just popping the key-value pair out of the dict is neater.